File: py/src/ag_ui_validate/report/pretty.py
"""Human-readable output. Pure string formatting - the CLI decides where it
goes and whether a TTY wants color. Mirrors src/report/pretty.ts.
"""

from __future__ import annotations

from typing import Dict, List

from ..rules.catalog import RULES
from ..types import Diagnostic, Report

_SYMBOL: Dict[str, str] = {"error": "✖", "warning": "⚠", "info": "ℹ"}
_SGR: Dict[str, str] = {"error": "31", "warning": "33", "info": "36"}
_SEVERITY_RANK: Dict[str, int] = {"error": 0, "warning": 1, "info": 2}
_SAMPLE_INDEXES = 3


def _paint(code: str, s: str, on: bool) -> str:
    return f"\x1b[{code}m{s}\x1b[0m" if on else s


def format_diagnostic_line(d: Diagnostic, color: bool) -> str:
    where = f"event {d.event_index}" if d.event_index >= 0 else "—"
    head = _paint(_SGR[d.severity], f"{_SYMBOL[d.severity]} {d.rule}", color)
    meta = _paint("2", f"{d.severity.ljust(7)} {where.ljust(10)}", color)
    cite = _paint("2", f"  ↳ {d.spec_url}", color)
    return f"{head}  {meta} {d.message}\n{cite}"


def _count(n: int, noun: str) -> str:
    return f"{n} {noun}{'' if n == 1 else 's'}"


def format_grouped_diagnostics(diagnostics: List[Diagnostic], color: bool) -> str:
    """One line per rule instead of one per occurrence - for large streams
    where the same violation repeats. Totals stay honest in the summary;
    this only changes what is listed."""
    groups: Dict[str, List[Diagnostic]] = {}
    for d in diagnostics:
        groups.setdefault(d.rule, []).append(d)

    sorted_groups = sorted(
        groups.values(), key=lambda g: (_SEVERITY_RANK[g[0].severity], g[0].rule)
    )

    lines: List[str] = []
    for group in sorted_groups:
        first = group[0]
        catalog = RULES.get(first.rule)
        title = catalog.title if catalog is not None else first.message
        indexes = [d.event_index for d in group if d.event_index >= 0]
        if not indexes:
            where = "stream-level"
        else:
            sample = ", ".join(str(i) for i in indexes[:_SAMPLE_INDEXES])
            rest = len(indexes) - _SAMPLE_INDEXES
            where = f"events {sample}{f' (+{rest} more)' if rest > 0 else ''}"
        head = _paint(_SGR[first.severity], f"{_SYMBOL[first.severity]} {first.rule}", color)
        meta = _paint("2", f"{first.severity.ljust(7)} ×{len(group)}", color)
        cite = _paint("2", f"  ↳ {first.spec_url}", color)
        lines.append(f"{head}  {meta}  {title} — {where}")
        lines.append(cite)
    return "\n".join(lines)


def format_report_summary(report: Report, color: bool) -> str:
    errors, warnings, info = report.summary.errors, report.summary.warnings, report.summary.info
    lines: List[str] = []

    if errors + warnings + info == 0:
        lines.append(
            _paint("32", f"✔ no conformance violations across {_count(report.event_count, 'event')}", color)
        )
    else:
        lines.append(
            f"{_count(errors, 'error')}, {_count(warnings, 'warning')}, {info} info across "
            f"{_count(report.event_count, 'event')}"
        )

    features = list(report.features.items())
    exercised = [f for f, s in features if s == "exercised"]
    suffix = f": {', '.join(exercised)}" if exercised else ""
    lines.append(f"{len(exercised)} of {len(features)} AG-UI features exercised{suffix}")

    if report.skipped:
        lines.append(f"{_count(len(report.skipped), 'rule')} not evaluated:")
        for s in report.skipped:
            lines.append(_paint("2", f"  – {s.rule}: {s.reason}", color))

    for e in report.internal_errors:
        lines.append(_paint("31", f"! internal validator error: {e}", color))

    return "\n".join(lines)

File: src/report/pretty.ts
// Human-readable output. Pure string formatting — the CLI decides where it
// goes and whether a TTY wants color.
import { RULES } from "../rules/catalog.js"
import type { Diagnostic, Report, Severity } from "../types.js"

export interface PrettyOptions {
  color: boolean
}

const SYMBOL: Record<Severity, string> = { error: "✖", warning: "⚠", info: "ℹ" }
const SGR: Record<Severity, string> = { error: "31", warning: "33", info: "36" }

function paint(code: string, s: string, on: boolean): string {
  return on ? `\x1b[${code}m${s}\x1b[0m` : s
}

export function formatDiagnosticLine(d: Diagnostic, opts: PrettyOptions): string {
  const where = d.eventIndex >= 0 ? `event ${d.eventIndex}` : "—"
  const head = paint(SGR[d.severity], `${SYMBOL[d.severity]} ${d.rule}`, opts.color)
  const meta = paint("2", `${d.severity.padEnd(7)} ${where.padEnd(10)}`, opts.color)
  const cite = paint("2", `  ↳ ${d.specUrl}`, opts.color)
  return `${head}  ${meta} ${d.message}\n${cite}`
}

function count(n: number, noun: string): string {
  return `${n} ${noun}${n === 1 ? "" : "s"}`
}

const SEVERITY_RANK: Record<Severity, number> = { error: 0, warning: 1, info: 2 }
const SAMPLE_INDEXES = 3

/**
 * One line per rule instead of one per occurrence — for large streams where
 * the same violation repeats. Totals stay honest in the summary; this only
 * changes what is listed.
 */
export function formatGroupedDiagnostics(diagnostics: Diagnostic[], opts: PrettyOptions): string {
  const groups = new Map<string, Diagnostic[]>()
  for (const d of diagnostics) {
    const list = groups.get(d.rule)
    if (list === undefined) groups.set(d.rule, [d])
    else list.push(d)
  }

  const sorted = [...groups.values()].sort((a, b) => {
    const bySeverity = SEVERITY_RANK[a[0]!.severity] - SEVERITY_RANK[b[0]!.severity]
    return bySeverity !== 0 ? bySeverity : a[0]!.rule.localeCompare(b[0]!.rule)
  })

  const lines: string[] = []
  for (const group of sorted) {
    const first = group[0]!
    const title = RULES.get(first.rule)?.title ?? first.message
    const indexes = group.map((d) => d.eventIndex).filter((i) => i >= 0)
    let where: string
    if (indexes.length === 0) {
      where = "stream-level"
    } else {
      const sample = indexes.slice(0, SAMPLE_INDEXES).join(", ")
      const rest = indexes.length - SAMPLE_INDEXES
      where = `events ${sample}${rest > 0 ? ` (+${rest} more)` : ""}`
    }
    const head = paint(SGR[first.severity], `${SYMBOL[first.severity]} ${first.rule}`, opts.color)
    const meta = paint("2", `${first.severity.padEnd(7)} ×${group.length}`, opts.color)
    const cite = paint("2", `  ↳ ${first.specUrl}`, opts.color)
    lines.push(`${head}  ${meta}  ${title} — ${where}`, cite)
  }
  return lines.join("\n")
}

export function formatReportSummary(report: Report, opts: PrettyOptions): string {
  const { errors, warnings, info } = report.summary
  const lines: string[] = []

  if (errors + warnings + info === 0) {
    lines.push(
      paint("32", `✔ no conformance violations across ${count(report.eventCount, "event")}`, opts.color),
    )
  } else {
    lines.push(
      `${count(errors, "error")}, ${count(warnings, "warning")}, ${info} info across ${count(report.eventCount, "event")}`,
    )
  }

  const features = Object.entries(report.features)
  const exercised = features.filter(([, s]) => s === "exercised").map(([f]) => f)
  const suffix = exercised.length > 0 ? `: ${exercised.join(", ")}` : ""
  lines.push(`${exercised.length} of ${features.length} AG-UI features exercised${suffix}`)

  if (report.skipped.length > 0) {
    lines.push(`${count(report.skipped.length, "rule")} not evaluated:`)
    for (const s of report.skipped) {
      lines.push(paint("2", `  – ${s.rule}: ${s.reason}`, opts.color))
    }
  }

  for (const e of report.internalErrors) {
    lines.push(paint("31", `! internal validator error: ${e}`, opts.color))
  }

  return lines.join("\n")
}
